src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -207,10 +210,72 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
                 }
             }
 
+            // We may get a flood of diagnostic notificaitons on the BG.  In order to not overwhelm
+            // the UI thread with work to do, and to avoid excess allocations, we batch up these
+            // notifications, and attempt to process them all at once on the UI thread every 50ms
+            // or so.
+            //
+            // Because we're batching, we can optimize things such that we only store two pieces
+            // of data per provider.  Specifically, if they were removing all diagnostics, and
+            // per-document, the latest diagnostics we've created for it.
+            private static readonly ObjectPool<ProviderIdToBatchedUpdates> s_providerPool = new ObjectPool<ProviderIdToBatchedUpdates>(() => new  ProviderIdToBatchedUpdates());
+            private static readonly ObjectPool<DocumentToUpdateArgs> s_documentPool = new ObjectPool<DocumentToUpdateArgs>(() => new DocumentToUpdateArgs());
+
+            private readonly object _gate = new object();
+            private ProviderIdToBatchedUpdates _idToBatchedUpdates = s_providerPool.Allocate();
+            private Task _updateTask = null;
+
             private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             {
                 this.AssertIsBackground();
-                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
+                if (_disposed)
+                {
+                    return;
+                }
+
+                lock (_gate)
+                {
+                    if (!_idToBatchedUpdates.TryGetValue(e.Id, out var batchedUpdates))
+                    {
+                        batchedUpdates = (removeArgs: null, createArgs: s_documentPool.Allocate());
+                    }
+
+                    switch (e.Kind)
+                    {
+                    case DiagnosticsUpdatedKind.DiagnosticsRemoved:
+                        // we're being told about diagnostics going away because a document/project
+                        // was removed.  This supercedes all previous removes and creates for this
+                        // id.
+                        batchedUpdates.removeArgs = e;
+
+                        // Clear all all creates for this provider. 
+                        batchedUpdates.createArgs.Clear();
+                        break;
+                    case DiagnosticsUpdatedKind.DiagnosticsCreated:
+                        // We're creating diagnostics. This supercedes all existing creations for
+                        // this id.  However, any existing removal for this id will still happen.
+                        var document = e.Solution.GetDocument(e.DocumentId);
+                        batchedUpdates.createArgs[document] = e;
```

Comment:
❓ Why not use `DocumentId` as the key here? The use of `Document` means entire solutions could end up rooted in the work queue.

Reply:
it's rooted no matter what.  e holds onto a solution.  So Document was more natural here, without changing what is rooted.